docs/hibernating-clusters.md

```diff
@@ -0,0 +1,121 @@
+# Hibernating Clusters
+
+## Overview
+
+Starting with OpenShift version 4.4.8, clusters can be stopped and started by simply shutting
+down their machines and starting them back up. The only additional requirement to bring a cluster
+back up from a stopped state is that CSRs be approved for the cluster's nodes when certificates
+have expired since the cluster was active.
+
+Hive can automate the process of stopping/starting clusters via its API by allowing the user to
+set a the desired state of the cluster in the ClusterDeployment spec. Both API and controller
+changes are required to support this feature.
+
+## API Changes
+
+The ClusterDeploymentSpec should allow setting whether machines are in a running state or in
+a hibernating state.
+
+```go
+type ClusterDeploymentSpec struct {
+  // ... other fields
+
+  // Hibernating indicates whether a cluster should be hibernating. Default is false.
+  Hibernating  `json:"hibernating,omitempty"`
+}
+```
+
+On the status side, a cluster's state is reflected by the following condition types:
+
+```go
+const (
+  // ... other constants
+
+  // ClusterStartingConditionType is set when the ClusterDeployment is transitioning
+  // from a hibernating state to a running state.
+  ClusterStartingConditionType ClusterDeploymentConditionType = "ClusterStarting"
+
+  // ClusterRunningConditionType is set when the ClusterDeployment is in a running and
+  // ready state.
+  ClusterRunningConditionType ClusterDeploymentConditionType = "ClusterRunning"
+
+  // ClusterStoppingConditionType is set when the ClusterDeployment is transitioning
+  // from a running state to a hibernating state.
+  ClusterStoppingConditionType ClusterDeploymentConditionType = "ClusterStopping"
+
+  // ClusterHibernatingConditionType is set when the ClusterDeployment has transitioned
+  // to a hibernating state.
+  ClusterHibernatingConditionType ClusterDeploymentConditionType = "ClusterHibernating"
+
+  // HibernationNotSupportedConditionType is set when a ClusterDeployment's hibernating
+  // property is set to true, but the cluster has an OpenShift version that doesn't support hibernating
+  // (< 4.4.8).
+  HibernationNotSupportedConditionType ClusterDeploymentConditionType = "HibernationNotSupported"
+)
+```
+
+## Controller Changes
+
+### Cluster Hibernation Controller
+The cluster hibernation controller is a new controller that will watch ClusterDeployments and ensure that the target cluster's
+machines reflect the hibernating state specified in the ClusterDeployment spec.
+
+High level flow of controller logic:
+
+![hibernation_controller](hibernation_controller.png)
```

Comment:
Once a cluster is hibernating, should the controller repeatedly verify that the machines are still stopped? In other words, if a user manually starts a machine while the cluster is hibernating, should Hive override that and stop the machine again?